xsnap/documentation/xsnap-worker.md

```diff
@@ -0,0 +1,52 @@
+
+Once started, the process listens on file descriptor 3, and will write to file descriptor 4. The process will perform a blocking read on fd3 until a complete netstring is received. The first character of the body of this netstring indicates what command to execute, with the remainder of the body as the command's payload. The commands are:
+
+* `R` (isReady): writes `netstring(".")` (the body is a single period) to fd4, to acknowledge that the worker is running
+* `e` (evaluate): evaluate the body in the top-level JS environment
+* `?` (command): feed the body (as a JS `String`) to the registered `handleCommand(body)` handler
+  * for both `e` and `?`, execution continues until both the `setImmediate` and the ready-promise-callback queues are empty (the vat is "quiescent")
+  * if evaluation/`handleCommand()` throws an error (and the evaluated code does not catch it), the worker writes `!${toString(err)}` to fd4
+  * if successful, the result should be an ArrayBuffer, or an object with a `.result` property that is an ArrayBuffer
+    * anything else will yield an empty response string
+  * the worker writes a netstring with the following body to fd4:
+    * `.${meterObj}\1${result}`
+    * the `.` prefix indicates success
+    * "meterObj" is a JSON-parseable record, with string keys and numeric values, currently containing `{ currentHeapCount, compute, allocate, timestamps }`
+      * `currentHeapCount` is the number of bytes allocated by the JS engine (`fxGetCurrentHeapCount()`), it never goes down
+      * `compute` is the number of computrons used during the evaluation/`handleCommand()` (`meterIndex)`
+      * `allocate` is `the->allocatedSpace`
+      * `timestamps` is an array of numbers (fractional seconds since epoch):
+        * the first is the time at which the `e` or `?` command was received by the worker (recorded just after the fd3 netstring is parsed)
+        * followed by a pair for each `issueCommand` sent to the parent: the first is the time just before the `issueCommand` is written to fd4, the second is the time just after the response is received on fd3
+        * the last is the time just before the `.${meterObj}\1${result}` body is serialized, immediately before it is written back to the parent on fd4 to complete the delivery
+      * only the first 100 such timestamps are reported; any later ones are omitted
+    * the meterObj is separated from the result by a 0x01 byte (i.e. U+0001 if the body is parsed as UTF-8)
+    * the `result` field is the ArrayBuffer
+* `s` (run script): the body is treated as the filename of a program to run (`xsRunProgramFile`)
+* `m` (load module): the body is treated as the name of a module to load (`xsRunModuleFile`). The module must already be defined, perhaps pre-compiled into the `xsnap` executable.
```

Comment:
The m and s features are probably best used via  xsnap.c . We copied them over and haven't deleted them, but we probably should. Maybe mark them deprecated?